File: JavaScript/056 Merge Intervals.js
// Given a collection of intervals, merge all overlapping intervals.

// Example 1:

// Input: [[1,3],[2,6],[8,10],[15,18]]
// Output: [[1,6],[8,10],[15,18]]
// Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
// Example 2:

// Input: [[1,4],[4,5]]
// Output: [[1,5]]
// Explanation: Intervals [1,4] and [4,5] are considerred overlapping.


/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
var merge = function (intervals) {
    if (intervals.length === 0) return [];

    intervals.sort(function (i, j) {
        return i.start - j.start;
    });
    let result = [intervals[0]];
    for (let i = 1; i < intervals.length; i++) {
        if (intervals[i].start > result[result.length - 1].end) {
            result.push(intervals[i]);
        } else {
            result[result.length - 1].end = Math.max(intervals[i].end, result[result.length - 1].end);
        }
    }
    return result;
};

// test
function Interval(start, end) {
    this.start = start;
    this.end = end;
}

let arr = [
    new Interval(1, 3),
    new Interval(2, 6),
    new Interval(8, 10),
    new Interval(15, 18)
];
console.log(merge(arr));
arr = [
    new Interval(1, 4),
    new Interval(4, 5),
];
console.log(merge(arr));

File: Python/056 Merge Intervals.py
# Given a collection of intervals, merge all overlapping intervals.

# Example 1:

# Input: [[1,3],[2,6],[8,10],[15,18]]
# Output: [[1,6],[8,10],[15,18]]
# Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
# Example 2:

# Input: [[1,4],[4,5]]
# Output: [[1,5]]
# Explanation: Intervals [1,4] and [4,5] are considerred overlapping.


# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if len(intervals) == 0:
            return []

        intervals.sort(lambda i, j: i.start - j.start)
        result = [intervals[0]]
        for i in range(1, len(intervals)):
            if intervals[i].start > result[-1].end:
                result.append(intervals[i])
            else:
                result[-1].end = max(intervals[i].end, result[-1].end)
        return result

# test
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

arr = [
    Interval(1, 3),
    Interval(2, 6),
    Interval(8, 10),
    Interval(15, 18)
]
s = Solution()
print(s.merge(arr))

